### PR TITLE
Performance

### PR DESCRIPTION
**Pull request by caseyvangroll**
Couldn't sleep so checked out Regent and had a super fun time.
Summary of changes:
- Replace String.replace with String.slice
- Trim unnecessary matches from end of regexes
- Extract regexes to be re-used
- Optimize xor evaluation

image

### CODE REVIEW

**Comment by aLittleTooQuiet** [comment]
Wow, impressive work!

**Review by eisenivan — COMMENTED** [review]
baller!

File: src/fn.js [hunk 1]
```diff
@@ -31,22 +30,18 @@ export default (id, custom) => (
       regex,
       some,
       typeOf,
+      ...custom,
```
Comment (eisenivan) [inline src/fn.js]: I think Object.assign might be faster than object spread. I haven't run the tests agains it, but this exists...

https://www.measurethat.net/Benchmarks/Show/2978/0/lodash-merge-vs-objectassign-vs-spread

Not a big deal, but if we're really looking to squeeze some speed out of here it's a thing we could look into.

**Review by eisenivan — COMMENTED** [review]
awesome work

**Review by eisenivan — APPROVED** [review]

**Review by aLittleTooQuiet — APPROVED** [review]

File: src/fn.js [hunk 2]
```diff
@@ -31,22 +30,18 @@ export default (id, custom) => (
...
 
     try {
-      if (id.indexOf('!') === 0) {
-        f = id.replace('!', '');
-        result = !assign({}, fn, custom)[f](left, right, data, custom);
-      } else {
-        result = assign({}, fn, custom)[f](left, right, data, custom);
-      }
+      const exec = key => fn[key](left, right, data, custom);
+      result = /^!/.test(id) ? !exec(id.slice(1)) : exec(id);
```
Comment (jesselee34) [inline src/fn.js:39]: exec is an expensive operation. It surprises me that this line would improve performance. Are we sure this specifically improves perf?

File: src/fn.js (hunk 1 above)
Comment (caseyvangroll) [reply src/fn.js]: This website is super handy - I'm gonna change it here

File: src/fn.js (hunk 2 above)
Comment (jesselee34) [reply src/fn.js:39]: Oh, I miss-understood this code. I'm thinking of `eval`. Never mind. This is rad.

**Review by jesselee34 — COMMENTED** [review]
Looks good. I like it.

**Pull request merged**